internal/config/image/cuda_image.go

```diff
@@ -216,14 +229,53 @@ func (i CUDA) OnlyFullyQualifiedCDIDevices() bool {
 	return hasCDIdevice
 }
 
-// VisibleDevicesFromEnvVar returns the set of visible devices requested through
-// the NVIDIA_VISIBLE_DEVICES environment variable.
+// VisibleDevices returns a list of devices requested in the container image.
+// If volume mount requests are enabled these are returned if requested,
+// otherwise device requests through environment variables are considered.
+// In cases where environment variable requests required privileged containers,
+// such devices requests are ignored.
+func (i CUDA) VisibleDevices() []string {
+	// If enabled, try and get the device list from volume mounts first
+	if i.acceptDeviceListAsVolumeMounts {
+		volumeMountDeviceRequests := i.visibleDevicesFromMounts()
+		if len(volumeMountDeviceRequests) > 0 {
+			return volumeMountDeviceRequests
+		}
+	}
+
+	// Get the Fallback to reading from the environment variable if privileges are correct
+	envVarDeviceRequests := i.VisibleDevicesFromEnvVar()
+	if len(envVarDeviceRequests) == 0 {
+		return nil
+	}
+
+	// If the container is privileged, or environment variable requests are
+	// allowed for unprivileged containers, these devices are returned.
+	if i.isPrivileged || i.acceptEnvvarUnprivileged {
+		return envVarDeviceRequests
+	}
+
+	// We log a warning if we are ignoring the environment variable requests.
+	i.logger.Warningf("Ignoring devices specified in NVIDIA_VISIBLE_DEVICES in unprivileged container")
```

Comment:
technically this could be checking other envvars. I'm ok to leave this as is though.

Reply:
You are right, because if the previous check is not fulfilled, we will land here, and if this is a SWARM use case for example we would also be ignoring `preferredVisibleDeviceEnvVars`

Reply:
(strictly speaking we should factor this into a function and use that from `VisibleDevicesFromEnvVar`, but I don't want to waste cycles on this). Let's do this as a follow-up. Please create an issue to track.

Reply:
I'll create an issue once we merge the PR, so I can point to places in the code in the issue

Reply:
how about

```go
// visibleEnvVars returns the environment variables that are used to determine device visibility.
// It returns the preferred environment variables that are set, or NVIDIA_VISIBLE_DEVICES if none are set.
func (i CUDA) visibleEnvVars() []string {
	var envVars []string
	for _, envVar := range i.preferredVisibleDeviceEnvVars {
		if i.HasEnvvar(envVar) {
			envVars = append(envVars, envVar)
		}
	}
	if len(envVars) == 0 {
		envVars = append(envVars, EnvVarNvidiaVisibleDevices)
	}
	return envVars
}
```

(I have tested it and it works for both the logging use case and the `VisibleDevicesFromEnvVar` use case